adapters/zerolog/zerolog.go

```diff
@@ -0,0 +1,286 @@
+// "Set".
+//
+// An API token with "ingest" permission is sufficient enough.
+//
+// A hook needs to be closed properly to make sure all logs are sent by calling
+// [Hook.Close].
+func New(opts ...writerOption) (*WriteCloser, error) {
+	w := &WriteCloser{
+		levels:        make(map[zerolog.Level]struct{}),
+		ingestOptions: []ingest.Option{},
+		clientOptions: []axiom.Option{},
+		byteCh:        make(chan []byte, defaultBatchSize),
+		closeCh:       make(chan struct{}),
+		ctx:           context.Background(),
+	}
+
+	// Apply supplied options.
+	for _, option := range opts {
+		if option == nil {
+			continue
+		}
+		option.apply(w)
+	}
+
+	if len(w.levels) == 0 {
+		for _, level := range []zerolog.Level{zerolog.InfoLevel, zerolog.WarnLevel, zerolog.ErrorLevel, zerolog.FatalLevel, zerolog.PanicLevel} {
+			w.levels[level] = struct{}{}
+		}
+	}
+
+	// Create client, if not set.
+	if w.client == nil {
+		var err error
+		if w.client, err = axiom.NewClient(w.clientOptions...); err != nil {
+			return nil, err
+		}
+	}
+
+	// When the dataset name is not set, use "AXIOM_DATASET".
+	if w.datasetName == "" {
+		w.datasetName = os.Getenv("AXIOM_DATASET")
+		if w.datasetName == "" {
+			return nil, ErrMissingDatasetName
+		}
+	}
+
+	go w.runBackgroundJob()
+	return w, nil
+}
+
+func (w *WriteCloser) runBackgroundJob() {
+	var (
+		counter = 0
+		buffer  = &bytes.Buffer{}
+		t       = time.NewTicker(flushInterval)
+		encoder = axiom.ZstdEncoder()
+	)
+
+	flush := func() error {
+		defer func() {
+			counter = 0
+			t.Reset(flushInterval)
+			buffer.Reset()
+		}()
+
+		if buffer.Len() == 0 {
+			return nil
+		}
+
+		r, err := encoder(buffer)
+		if err != nil {
+			return err
+		}
+
+		res, err := w.client.Ingest(w.ctx, w.datasetName, r, axiom.NDJSON, axiom.Zstd, w.ingestOptions...)
+		if err != nil {
+			return fmt.Errorf("failed to ingest events: %w", err)
+		}
+		if res.Failed > 0 {
+			return fmt.Errorf("failed to ingest events: %w", err)
+		}
+		if res.Failed > 0 {
+			logger.Printf("event(s) [%v] at %s failed to ingest: %s\n", res.Failed, res.Failures[0].Timestamp, res.Failures[0].Error)
+			return nil
+		}
+		return nil
+	}
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			if err := flush(); err != nil {
+				logger.Printf("failed to ingest events: %s\n", err)
+			}
+			close(w.closeCh)
+			return
+		case data, ok := <-w.byteCh:
+			if !ok {
+				if err := flush(); err != nil {
+					logger.Printf("failed to ingest events: %s\n", err)
+				}
+				close(w.closeCh)
+				return
+			}
+			if len(data) == 0 {
+				continue
+			}
+
+			counter++
+
+			lvlStr, err := jsonparser.GetUnsafeString(data, zerolog.LevelFieldName)
+			if err != nil {
+				logger.Printf("failed to retrieve level field name from data: %s\n", err)
+				continue
+			}
+
+			lvl, err := zerolog.ParseLevel(lvlStr)
+			if err != nil {
+				logger.Printf("failed to parse level: %s\n", err)
+				continue
+			}
+
+			if _, enabled := w.levels[lvl]; !enabled {
+				continue
+			}
```

Comment:
I wonder: Is there no way to let `zerolog` control which levels get send to our adapter?